smartsim/entity/files.py

```diff
@@ -130,3 +130,122 @@ def _check_path(self, file_path):
         if path.isdir(full_path):
             return full_path
         raise SSConfigError(f"File or Directory {file_path} not found")
+
+
+class TaggedFilesHierarchy:
+    """The TaggedFilesHierarchy class maintains a list of files and a list of
+    child TaggedFilesHierarchy instances. Each instance describes how a
+    directory structure of tagged files.
+
+    TaggedFilesHierarchy represents a directory containing potentially tagged
+    files and subdirectories.
+
+    TaggedFilesHierarchy.base is the directory path from the the root of the
+    file structure
+    E.g.
+    TaggedFilesHierarchy.base = ""            => ./
+    TaggedFilesHierarchy.base = "dir_a"       => ./dir_a
+    TaggedFilesHierarchy.base = "dir_a/dir_b" => ./dir_a/dir_b
+
+    TaggedFilesHierarchy.files is a collection of paths to files that need
+    to be copied to directory that the TaggedFilesHierarchy represents
+
+    TaggedFilesHierarchy.dirs is a collection of child TaggedFilesHierarchy,
+    each with its own files and dirs, representing a directory that needs to
+    be created within the current directory represented by the current
+    TaggedFilesHierarchy
+
+    By performing a depth first search over the entire hierarchy starting at
+    the root (TaggedFilesHierarchy.base = ""), one could reconstruct the
+    tagged file directory structure with relative ease at any location deemed
+    necessary.
+    """
+
+    def __init__(self, base=""):
+        """Initialize a TaggedFilesHierarchy
+
+        :param base: pathlike string specifing the generated directory
+                     files are located
+        :type base: str, optional
+        """
+        self.base = base
+        self.files = set()
+        self.dirs = set()
+
+    @classmethod
+    def from_list_paths(cls, path_list, dir_contents_to_base=False):
+        """Given a list of absolute paths to files and dirs, create and return
+        a TaggedFilesHierarchy instance representing the file hierarchy of
+        tagged files. All files in the path list will be placed in the base of
+        the file hierarchy.
+
+        :param path_list: list of absolute paths to tagged files or dirs
+                          containing tagged files
+        :type path_list: list[str]
+        :param dir_contents_to_base: When a top level dir is encountered, if
+                                     this value is truthy, files in the dir are
```

Comment:
truthy?

Reply:
Vocab is used mostly in the javascript world but I've seen it used with python as well. Means that the value evaluates to true (so `True`, `1`, and `"hello"` are all truthy). Similarly, falsy means evaluates to false (`False`, `0`, `""`, etc). I'm inclined to keep it as it doesn't really matter if a boolean value is passed

Reply:
Learn something every day. I use the term ``boolable`` which might not be a word.

I was wondering what the mypy type would be. Interesting discussion about that here: https://stackoverflow.com/questions/51654434/the-type-of-a-variable-that-can-be-implicitly-checked-for-zeroness-or-emptine

Reply:
The type sig is ``bool`` so I don't think people are getting confused. 

Consider resolved.